# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/tests/browser/fission/browser_take_focus.js
# Commit: b3216f1cfd1f
# Full Hash: b3216f1cfd1f295a707bb089132f86c1e0770c8d
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2021-02-02 03:35:00
# Regressor Bug: 1688972
# File Overlap Count: 1
# Description:
#   Bug 1688972 - Make FocusedChild work across document/process boundaries. r=Jamie
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D103102
# ==============================================================================

diff -r 0a6d78effab1 -r b3216f1cfd1f accessible/tests/browser/fission/browser_take_focus.js
--- a/accessible/tests/browser/fission/browser_take_focus.js	Mon Feb 01 20:09:55 2021 +0100
+++ b/accessible/tests/browser/fission/browser_take_focus.js	Mon Feb 01 17:54:23 2021 +0000
@@ -11,9 +11,15 @@
 );
 
 addAccessibleTask(
-  `<input id="textbox" value="hello"/>`,
+  `<div role="group"><input id="textbox" value="hello"/></div>`,
   async function(browser, iframeDocAcc, contentDocAcc) {
     const textbox = findAccessibleChildByID(iframeDocAcc, "textbox");
+    const iframe = findAccessibleChildByID(contentDocAcc, "default-iframe-id");
+    const iframeDoc = findAccessibleChildByID(
+      contentDocAcc,
+      "default-iframe-body-id"
+    );
+
     testStates(textbox, STATE_FOCUSABLE, 0, STATE_FOCUSED);
 
     let onFocus = waitForEvent(EVENT_FOCUS, textbox);
@@ -21,6 +27,40 @@
     await onFocus;
 
     testStates(textbox, STATE_FOCUSABLE | STATE_FOCUSED, 0);
+
+    if (Services.appinfo.OS == "WINNT") {
+      // focusedChild XPCOM method not implemented in windows.
+      return;
+    }
+
+    is(
+      getAccessibleDOMNodeID(contentDocAcc.focusedChild),
+      "textbox",
+      "correct focusedChild from top doc"
+    );
+
+    is(
+      getAccessibleDOMNodeID(iframeDocAcc.focusedChild),
+      "textbox",
+      "correct focusedChild from child doc"
+    );
+
+    ok(!iframe.focusedChild, "correct focusedChild from iframe (null)");
+
+    onFocus = waitForEvent(EVENT_FOCUS, iframeDoc);
+    iframeDoc.takeFocus();
+    await onFocus;
+
+    is(
+      getAccessibleDOMNodeID(contentDocAcc.focusedChild),
+      "default-iframe-body-id",
+      "correct focusedChild of child doc from top doc"
+    );
+    is(
+      getAccessibleDOMNodeID(iframe.focusedChild),
+      "default-iframe-body-id",
+      "correct focusedChild of child doc from iframe"
+    );
   },
   { topLevel: false, iframe: true, remoteIframe: true }
 );
